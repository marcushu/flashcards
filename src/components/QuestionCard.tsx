import { useState, useEffect, useContext } from 'react';
import UserContext from '../UserContext';
import { isNil } from 'lodash'

interface Question {
  question?: string,
  answer?: string,
}

interface QuestionCardProps {
  randomQuestion: Question,
  setshowModal: (newState: boolean) => void
  nextQuestion: () => void
  deleteQuestion: (question: string) => void
}

const QuestionCard = ({ randomQuestion, setshowModal, nextQuestion, deleteQuestion }: QuestionCardProps) => {
  const [hideQuestion, sethideQuestion] = useState(true);
  const user = useContext(UserContext);

  useEffect(() => {
    sethideQuestion(true);
  }, [randomQuestion]);

  const newQuestion = () => {
    sethideQuestion(true);
    nextQuestion();
  }

  const deleteIt = () => {
    if (!isNil(randomQuestion.question)) {
      deleteQuestion(randomQuestion.question);
      newQuestion();
    }
  }

  return (
    <div id="questionCard">
      {randomQuestion.question
        ?
        <>
          <div id="cardHeader">
            <div className="cardHead">
              {user.userStatus !== 'GUEST' &&
                <>
                  <button className="addDelete" style={{ width: "30px", color: "red" }} onClick={deleteIt}>
                    -
                  </button>
                  <button className="addDelete" style={{ width: "30px", color: "rgb(123 123 123)" }} onClick={() => setshowModal(true)}>
                    +
                  </button>
                </>}
              &nbsp;
              {randomQuestion.question}
            </div>
          </div>
          <div id="cardBody" style={{height: "auto", paddingBottom: "15px"}}>
            {hideQuestion
              ?
              <div>
                <button id="questionMask"
                  onClick={() => sethideQuestion(false)}
                  style={{ visibility: !hideQuestion ? "hidden" : "visible" }}>?</button>
              </div>
              :
              <div style={{ visibility: hideQuestion ? "hidden" : "visible" }}>
                {randomQuestion.answer}
              </div>
            }
          </div>
          <div id="cardFooter">
            <button className="questionButtons" style={{ width: "100%" }} onClick={newQuestion}>
              &gt;&gt;
            </button>
          </div>
        </>
        :
        <div id="cardBody">
          <p style={{ fontStyle: "italic" }}>
            Questions will appear here when a topic is selected
          </p>
        </div>
      }
    </div>
  )
}

export default QuestionCard;