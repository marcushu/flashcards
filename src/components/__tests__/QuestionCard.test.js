import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import QuestionCard from '../QuestionCard';

describe("question card", () => {
    const mockNextQuestion = jest.fn();
    const mockHideMe = jest.fn();
    const mockQuestion  = {
        question: "fake question",
        answer: "fake answer"
    }

    it("renders", () => {
        const { asFragment } = render(
            <QuestionCard nextQuestion={mockNextQuestion}
                setshowModal={mockHideMe}
                randomQuestion={mockQuestion} />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("displays question", () => {
        const { getByText } = render(
            <QuestionCard nextQuestion={mockNextQuestion}
                setshowModal={mockHideMe}
                randomQuestion={mockQuestion} />
        );

        expect(getByText(/fake question/i)).toBeInTheDocument();
    });

    it("has a working 'next' button", () => {
        const { getByRole } = render(
            <QuestionCard nextQuestion={mockNextQuestion}
                setshowModal={mockHideMe}
                randomQuestion={mockQuestion} />
        );

        fireEvent.click(getByRole('button', {  name: />>/i}));
        
        expect(mockNextQuestion).toHaveBeenCalled();
    });

    it("hides a 'new' question", () => {
        const { getByRole } = render(
            <QuestionCard nextQuestion={mockNextQuestion}
                setshowModal={mockHideMe}
                randomQuestion={mockQuestion} />
        );
        expect(getByRole('button', {  name: /\?/i})).toHaveStyle("visibility: visible");
    });

    it("shows the answer when the button is clicked", () => {
        const { getByRole, getByText } = render(
            <QuestionCard nextQuestion={mockNextQuestion}
                setshowModal={mockHideMe}
                randomQuestion={mockQuestion} />
        );

        fireEvent.click(getByRole('button', {  name: /\?/i}));

        expect(getByText(  /fake answer/i  )).toHaveStyle("visibility: visible");       
    });
})