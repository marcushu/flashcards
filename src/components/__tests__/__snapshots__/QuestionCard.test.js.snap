// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`question card renders 1`] = `
<DocumentFragment>
  <div
    id="questionCard"
  >
    <div
      id="cardHeader"
    >
      <div
        class="cardHead"
      >
        <button
          class="addDelete"
          style="width: 30px; color: red;"
        >
          -
        </button>
        <button
          class="addDelete"
          style="width: 30px;"
        >
          +
        </button>
        Â fake question
      </div>
    </div>
    <div
      id="cardBody"
      style="height: auto; padding-bottom: 15px;"
    >
      <div>
        <button
          id="questionMask"
          style="visibility: visible;"
        >
          ?
        </button>
      </div>
    </div>
    <div
      id="cardFooter"
    >
      <button
        class="questionButtons"
        style="width: 100%;"
      >
        &gt;&gt;
      </button>
    </div>
  </div>
</DocumentFragment>
`;
